R/grouped_epi_archive.R

```diff
@@ -357,15 +360,15 @@ grouped_epi_archive =
               }
 
               return(
-                dplyr::group_by(as_of_df, dplyr::across(tidyselect::all_of(private$vars)),
+                dplyr::group_by(as_of_df, !!!syms(private$vars),
                                 .drop=private$drop) %>%
                   dplyr::group_modify(group_modify_fn,
                                       f = f, ...,
                                       ref_time_value = ref_time_value,
                                       new_col = new_col,
                                       .keep = TRUE)
               )
-            })
+            }) %>% rbindlist()
```

Comment:
Nice speedup!  Is this going to change the output class of our result?  This might need to have a `setDF` (but check to make sure we can't possibly be aliasing something, e.g., in the 1-reference-time case).

Reply:
I fixed the class aspect, but not sure about the aliasing. My attempt at investigating this using `lobstr::ref` to compare addresses of list of epi_dfs containing computed values and tibble of computed values after using rbindlist:

```
> data$DT

# Input data
geo_value sex time_value    version value
1:        al   M 2012-01-01 2012-01-01    25
2:        ca   M 2012-01-01 2012-01-01   100

> data %>%
  +   epix_slide(f = ~ 5, before = 0, ref_time_values = as.Date("2012-01-01")) %>% 
  +   ungroup()

# output from lobstr::ref(<list of epi_dfs containing computed values >, <tibble of computed values after using rbindlist>)
█ [1:0x559b7f330b08] <list> 
  └─█ [2:0x559b80ee4dd8] <tibble[,2]> 
  ├─time_value = [3:0x559b7f330b40] <date> 
  └─slide_value = █ [4:0x559b7f467918] <list> 
  └─[5:0x559b7f331160] <dbl> 
  
  █ [6:0x559b80ee37d8] <tibble[,2]> 
  ├─time_value = [7:0x559b7f4842f8] <date> 
  └─slide_value = █ [8:0x559b7f484288] <list> 
  └─[5:0x559b7f331160] 

# Output of epix_slide
# A tibble: 1 × 2
time_value slide_value
* <date>           <dbl>
  1 2012-01-01           5
```

Reply:
Does this maintain the speed benefits?  `group_by` could potentially be costly.  I'll try to double-check the no-aliasing part.

Reply:
> Does this maintain the speed benefits? group_by could potentially be costly.

Yes, it seems to. I was also wondering about the time for that group_by but it didn't seem to impact anything.

Reply:
Great!  Regarding aliasing:
- Actually, we don't need to avoid it; we are fine with out outputs aliasing stuff I think.
- I've updated the docs to reflect the above and some other mutation&aliasing details.  Please sanity check.  (I hope we can eventually simplify & make all these notes internal by providing a filter method & hiding away the DT from the user.)
- We already probably don't ever alias.
  - `all_versions = FALSE` gets input that has gone through a DT -> tibble conversion using `as_tibble` with no `setDT`, which should prevent DT/column aliasing
  - `all_versions = TRUE` has a manual step making sure we don't alias the `DT` (I'm actually not sure why now... maybe I was thinking we were going to be mutating like in `fill_through_version`) but might potentially somehow alias the columns (I'm not sure if this is possible... but since the aliasing case of `unique` didn't appear to be documented before, we probably should assume it could change to alias columns instead of the DT).

Reply:
Thanks for adding all of those comments, they look good.

I'll remove some pipes as a final change.

Reply:
Did removing pipes actually make a difference in your testing?  I just tried re-introducing magrittr pipes to what I thought would be the highest-traffic pipe operations, the `group_by` and `group_modify` inside the `FUN` arg to `lapply`, but don't see any real change.

Reply:
Also it'd be nice if you could maybe summarize in general what seemed like the highest-impact changes here or any other places you think might benefit significantly from optimization.

Reply:
I don't see that de-piping changed runtime. Theoretically, extra function calls will always take [more time](https://www.r-bloggers.com/2011/10/overhead-cost-of-a-function-call/), but the cost is just so low compared to everything else we're doing here.

On the other hand, not using pipes does make profiles easier to read. Using pipes somehow causes functions to be marked as not having a sensible parent call, so e.g. `group_modify` shows up in a spot on the profile that you don't expect and that doesn't match the code.